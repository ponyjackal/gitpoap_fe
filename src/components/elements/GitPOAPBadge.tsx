import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { TextLight, MidnightBlue } from '../../colors';

type Props = {
  imgUrl: string;
};

type HexProps = {
  size: number;
  borderSize: number;
};

const Hexagon = styled.div`
  clip-path: polygon(5% 25%, 50% 0, 95% 25%, 95% 75%, 50% 100%, 5% 75%);
`;

const HexOuterBorder = styled(Hexagon)<HexProps>`
  position: relative;
  --s: ${(props) => rem(props.size + 4 * props.borderSize)};
  width: var(--s);
  height: calc(var(--s) * 1);
  background-color: ${TextLight};
`;
const HexInnerBorder = styled(Hexagon)<HexProps>`
  --s: ${(props) => rem(props.size + 2 * props.borderSize)};
  position: absolute;
  top: ${(props) => rem(props.borderSize)};
  left: ${(props) => rem(props.borderSize)};

  width: var(--s);
  height: calc(var(--s) * 1);
  background: ${MidnightBlue};
`;

const HexBadge = styled(Hexagon)<{ url: string } & HexProps>`
  --s: ${(props) => rem(props.size)};
  position: absolute;
  top: ${(props) => rem(props.borderSize)};
  left: ${(props) => rem(props.borderSize)};

  width: var(--s);
  height: calc(var(--s) * 1);
  display: inline-block;
  font-size: initial; /* we reset the font-size if we want to add some content */
  background: url('${(props) => props.url}') no-repeat center center;
  background-size: cover;
`;

export const GitPOAPBadge = ({ imgUrl }: Props) => {
  const size = 200;
  const borderSize = 4;
  return (
    <HexOuterBorder size={size} borderSize={borderSize}>
      <HexInnerBorder size={size} borderSize={borderSize}>
        <HexBadge url={imgUrl} size={size} borderSize={borderSize} />
      </HexInnerBorder>
    </HexOuterBorder>
  );
};

/* Another option is to use SVG ~ https://stackoverflow.com/questions/67458234/mask-image-with-svg-shape-and-add-a-border*/
