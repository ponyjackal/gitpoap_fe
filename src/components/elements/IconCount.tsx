import styled from 'styled-components';
import { rem } from 'polished';
import { ExtraHover, ExtraPressed, TextGray, TextLight } from '../../colors';

type Props = {
  icon: React.ReactNode;
  href?: string;
  count: number;
};

const Count = styled.span`
  font-family: PT Mono;
  font-style: normal;
  font-weight: bold;
  font-size: ${rem(12)};
  line-height: ${rem(17)};
  letter-spacing: ${rem(0.5)};
  text-transform: uppercase;
  color: ${TextLight};
  margin-left: ${rem(6)};
`;

const Container = styled.a`
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;

  path {
    transition: 200ms fill ease;
    fill: ${TextGray};
  }

  &:hover {
    path {
      fill: ${ExtraHover};
    }
  }
  &:active {
    path {
      fill: ${ExtraPressed};
    }
  }
`;

export const IconCount = ({ icon, href, count }: Props) => {
  return (
    <Container href={href}>
      {icon}
      {count && <Count>{count}</Count>}
    </Container>
  );
};
