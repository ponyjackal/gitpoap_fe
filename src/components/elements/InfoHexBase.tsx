import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { BackgroundPanel } from '../../colors';

type Props = {
  children: React.ReactNode;
};

const StyledInfoHex = styled.div`
  display: inline-flex;
  flex-direction: column;
  min-width: ${rem(250)};
`;

const UpperTriangle = styled.div`
  clip-path: polygon(50% 0%, 0% 100%, 100% 100%);
  width: 100%;
  height: ${rem(70)};
  background-color: ${BackgroundPanel};
`;

const LowerTriangle = styled(UpperTriangle)`
  clip-path: polygon(50% 100%, 0 0, 100% 0);
`;

const Body = styled.div`
  background-color: ${BackgroundPanel};
  min-height: ${rem(100)};
`;

export const InfoHexBase = (props: Props) => {
  return (
    <StyledInfoHex>
      <UpperTriangle />
      <Body>{props.children}</Body>
      <LowerTriangle />
    </StyledInfoHex>
  );
};
