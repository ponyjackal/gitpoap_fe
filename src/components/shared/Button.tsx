import React from 'react';
import styled from 'styled-components';
import { Button as MButton } from '@mantine/core';
import { rem } from 'polished';
import { PrimaryBlue, TextAccent } from '../../colors';

type Props = {
  children: React.ReactNode;
  className?: string;
  onClick?: React.MouseEventHandler;
  disabled?: boolean;
  leftIcon?: React.ComponentProps<typeof MButton>['leftIcon'];
};

const StyledButton = styled(MButton)`
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  background-color: ${PrimaryBlue};
  font-family: 'PT Mono', monospace;
  font-weight: 700;
  color: white;
  text-transform: uppercase;
  font-size: ${rem(12)};
  line-height: ${rem(18)};
  letter-spacing: ${rem(2)};
  transition: 200ms background ease;
  border-radius: ${rem(6)};
  padding: ${rem(8)} ${rem(14)};
  height: auto;

  &:hover {
    box-shadow: none;
    background-color: ${TextAccent};
  }

  &:active {
    background-color: #20457e;
  }
`;

export const Button = ({ children, className, onClick, disabled, leftIcon }: Props) => {
  return (
    <StyledButton
      compact
      className={className}
      size="md"
      disabled={disabled}
      onClick={onClick}
      leftIcon={leftIcon}
    >
      {children}
    </StyledButton>
  );
};
