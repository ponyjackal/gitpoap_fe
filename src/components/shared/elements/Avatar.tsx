import styled from 'styled-components';
import { rem } from 'polished';
import ImageUI from 'next/image';

type Props = {
  quality?: number;
  size?: number;
  className?: string;
  src: string;
  useDefaultImageTag?: boolean;
};

const AvatarWrapper = styled.div<{ size: number }>`
  height: ${({ size }) => rem(size)};
  width: ${({ size }) => rem(size)};
  position: relative;
`;

const Image = styled(ImageUI)`
  border-radius: 50%;
`;

const DefaultImage = styled.img`
  border-radius: 50%;
  width: 100%;
  height: 100%;
`;

export const Avatar = ({ quality = 100, size = 80, className, src, useDefaultImageTag }: Props) => (
  <AvatarWrapper className={className} size={size}>
    {useDefaultImageTag ? (
      <DefaultImage src={src} alt="" />
    ) : (
      <Image src={src} quality={quality} alt="" width={size} height={size} />
    )}
  </AvatarWrapper>
);
