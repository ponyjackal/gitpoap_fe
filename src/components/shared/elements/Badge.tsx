import styled from 'styled-components';
import { rem } from 'polished';
import { BackgroundPanel3, TextLight } from '../../../colors';

export const Badge = styled.div`
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  padding: ${rem(0.5)} ${rem(4)};
  background: ${BackgroundPanel3};
  border-radius: ${rem(4)};

  font-family: PT Mono;
  font-style: normal;
  font-weight: bold;
  font-size: ${rem(11)};
  line-height: ${rem(18)};
  letter-spacing: ${rem(1.2)};
  text-transform: uppercase;
  color: ${TextLight};
`;
