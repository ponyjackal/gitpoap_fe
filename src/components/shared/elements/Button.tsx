import React from 'react';
import styled from 'styled-components';
import { Button as ButtonUI, SharedButtonProps, ButtonProps } from '@mantine/core';
import { rem } from 'polished';
import { PrimaryBlue, TextGray, ExtraHover, ExtraPressed, DarkGray } from '../../../colors';

type Props = SharedButtonProps & {
  className?: string;
  onClick?: React.MouseEventHandler;
  onMouseEnter?: React.MouseEventHandler;
  onMouseLeave?: React.MouseEventHandler;
  disabled?: boolean;
  variant?: 'filled' | 'outline';
};

const StyledButton = styled(ButtonUI)<ButtonProps<'button'>>`
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  font-family: 'PT Mono', monospace;
  font-weight: 700;
  text-transform: uppercase;
  font-size: ${rem(12)};
  line-height: ${rem(18)};
  letter-spacing: ${rem(2)};
  transition: 150ms background ease, 150ms color ease, 150ms border ease;
  border-radius: ${rem(6)};
  padding: ${rem(8)} ${rem(14)};
  height: fit-content;

  &.mantine-Button-filled {
    background-color: ${PrimaryBlue};
    border: none;
    color: white;
    &:hover:not(:disabled) {
      background-color: ${ExtraHover};
    }
    &:active:not(:disabled) {
      background-color: ${ExtraPressed};
    }
    &:disabled {
      background-color: ${DarkGray};
      .mantine-Button-label {
        color: ${TextGray};
      }
    }
    &.mantine-Button-loading {
      background-color: ${PrimaryBlue};
      .mantine-Button-label {
        color: white;
      }
    }
  }

  &.mantine-Button-outline {
    border: ${rem(2)} solid ${TextGray};
    color: white;
    background-color: transparent;
    padding: ${rem(6)} ${rem(12)}; /* Adjusted for border */
    &:hover:not(:disabled) {
      border-color: ${ExtraHover};
      color: ${ExtraHover};
    }
    &:active:not(:disabled) {
      border-color: ${ExtraPressed};
      color: ${ExtraPressed};
    }
    &:disabled {
      background-color: transparent;
      border-color: ${TextGray};
      .mantine-Button-label {
        color: ${TextGray};
      }
    }
    &.mantine-Button-loading {
      &:before {
        top: ${rem(-2)};
        left: ${rem(-2)};
        right: ${rem(-2)};
        bottom: ${rem(-2)};
      }
      .mantine-Button-label {
        color: white;
      }
    }
  }
`;

export const Button = (props: Props) => {
  return <StyledButton {...props}>{props.children}</StyledButton>;
};
