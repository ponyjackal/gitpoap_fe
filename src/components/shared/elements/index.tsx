export * from './Avatar';
export * from './Badge';
export * from './Button';
export * from './Checkbox';
export * from './ClaimCircle';
export * from './CollapsibleAddress';
export * from './CopyableText';
export * from './DateInput';
export * from './Divider';
export * from './GitPOAPBadge';
export * from './Header';
export * from './HexagonPath';
export * from './IconCount';
export * from './InfoHexBase';
export * from './Input';
export * from './InputWrapper';
export * from './Label';
export * from './Loader';
export * from './MultiSelect';
export * from './NavLink';
export * from './NumberInput';
export * from './OrgName';
export * from './Pagination';
export * from './POAPBadge';
export * from './Radio';
export * from './RelativeDate';
export * from './Select';
export * from './Share';
export * from './Skeletons';
export * from './Text';
export * from './TextArea';
export * from './Title';
export * from './TwitterShareButton';
