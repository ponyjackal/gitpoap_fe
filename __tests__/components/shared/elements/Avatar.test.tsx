import { render } from '@testing-library/react';
import 'jest-styled-components';
import { Avatar } from '../../../../src/components/shared/elements';

describe('Avatar', () => {
  it('renders an Avatar - default img tag', () => {
    const { container } = render(<Avatar src="https://test.com/image.png" useDefaultImageTag />);
    const avatar = container.firstChild;
    const avatarImg = container.querySelector('img');

    expect(avatar).toBeInTheDocument();
    expect(avatar).toMatchSnapshot();
    /* expect avatar image to have src attribute */
    expect(avatarImg).toHaveAttribute('src', 'https://test.com/image.png');
  });

  it('renders an Avatar - Next Image', () => {
    const { container } = render(<Avatar src="https://test.com/image.png" />);
    const avatar = container.firstChild;

    expect(avatar).toBeInTheDocument();
    expect(avatar).toMatchSnapshot();
  });
});
