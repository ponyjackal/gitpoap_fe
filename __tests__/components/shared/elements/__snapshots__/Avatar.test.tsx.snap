// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders an Avatar - Next Image 1`] = `
.c0 {
  height: 5rem;
  width: 5rem;
  position: relative;
}

.c1 {
  border-radius: 50%;
}

<div
  class="c0"
  size="80"
>
  <img
    alt=""
    class="c1"
    data-nimg="1"
    decoding="async"
    height="80"
    loading="lazy"
    src="/_next/image?url=https%3A%2F%2Ftest.com%2Fimage.png&w=256&q=100"
    srcset="/_next/image?url=https%3A%2F%2Ftest.com%2Fimage.png&w=96&q=100 1x, /_next/image?url=https%3A%2F%2Ftest.com%2Fimage.png&w=256&q=100 2x"
    style="color: transparent;"
    width="80"
  />
</div>
`;

exports[`Avatar renders an Avatar - default img tag 1`] = `
.c0 {
  height: 5rem;
  width: 5rem;
  position: relative;
}

.c1 {
  border-radius: 50%;
  width: 100%;
  height: 100%;
}

<div
  class="c0"
  size="80"
>
  <img
    alt=""
    class="c1"
    src="https://test.com/image.png"
  />
</div>
`;
